basement/Basement/Bits.hs

```diff
@@ -0,0 +1,486 @@
+-- |
+-- Module      : Basement.Bits
+-- License     : BSD-style
+-- Maintainer  : Haskell Foundation
+-- Stability   : experimental
+-- Portability : portable
+--
+
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Basement.Bits
+    ( Bit
+    , BitOps(..)
+    , PrimBitOps(..)
+
+    , Bits
+    , toBits
+    , allOne
+    ) where
+
+import Basement.Compat.Base
+import Basement.Compat.Natural
+import Basement.Numerical.Additive
+import Basement.Numerical.Subtractive
+import Basement.Numerical.Multiplicative
+import Basement.Types.OffsetSize
+import Basement.Types.Word128 (Word128)
+import qualified Basement.Types.Word128 as Word128
+import Basement.Types.Word256 (Word256)
+import qualified Basement.Types.Word256 as Word256
+import Basement.IntegralConv (wordToInt)
+import Basement.Nat
+
+import qualified Prelude
+import qualified Data.Bits as OldBits
+import Data.Maybe (fromMaybe)
+import Data.Proxy
+import GHC.Base hiding ((.))
+import GHC.Prim
+import GHC.Types
+import GHC.Word
+import GHC.Int
+
+-- | type level tag to represet a bit.
+--
+-- Mainly used for offsets and sizes
+type Bit = Bool
+
+-- | operation over finit bits
+class PrimBitOps bits where
```

Comment:
I don't think it make sense to call it PrimBitOpts, I'ld rather see FiniteBitsOps